### PR TITLE
Queue class with tests

### PR DESCRIPTION
**Pull request by sdweber422**

### CODE REVIEW

File: spec/queue.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import chai, { expect, assert } from 'chai'
+import chaiChange from 'chai-change'
+import Queue from '../src/queue'
+
+chai.use(chaiChange)
+
+describe('Queue', () => {
+  'use strict'
```
Comment (jrobcodes) [inline spec/queue.js]: Out of curiosity, why do you use `use strict` here, instead of at the top of the file?

File: spec/queue.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    myQueue = new Queue()
+  })
+
+  it('exists', () => {
+    expect(Queue).to.be.a('function')
+  })
+
+  context('#enqueue(element)', () => {
+    it('should add an element to the back of the queue', () => {
+      expect(() => myQueue.enqueue("foo")).to.alter(() => myQueue.length(), { from: 0, to: 1 })
```
Comment (jrobcodes) [inline spec/queue.js]: Consider adding some whitespace to make this more readable - I have a hard time processing the embedded functions with parameters with it condensed on one line like this.

File: spec/queue.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  let myQueue
+  beforeEach( () => {
+    myQueue = new Queue()
+  })
+
+  it('exists', () => {
+    expect(Queue).to.be.a('function')
+  })
+
+  context('#enqueue(element)', () => {
```
Comment (jrobcodes) [inline spec/queue.js]: There is a test missing that validates that the correct `element` was enqueued.

File: spec/queue.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    it('should add an element to the back of the queue', () => {
+      expect(() => myQueue.enqueue("foo")).to.alter(() => myQueue.length(), { from: 0, to: 1 })
+    })
+  })
+
+  context('#dequeue', () => {
+    it('should remove and return the front element in the queue', () => {
```
Comment (jrobcodes) [inline spec/queue.js]: Whenever I see "and" or "or" in a test description, I am suspicious that the test is trying to validate too many things.  I think the two tests here are: `it( 'should remove an element from the queue' )`, and `it( 'should remove the first element from the queue' )`.

File: spec/queue.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    })
+    it('should return null if queue is empty', () => {
+      expect(myQueue.dequeue()).to.eql(null)
+    })
+  })
+
+  context('#front()', () => {
+    it('should return the front element of the queue', () => {
+      myQueue.enqueue("foo")
+      expect(myQueue.front()).to.eql("foo")
```
Comment (jrobcodes) [inline spec/queue.js]: Just want to point out that this test could be misleading - since there's only one element in the queue under test, it's possible we have a bug that we wouldn't see, since there's only one element that could possibly be returned. I would consider adding another element, just to ensure the *correct* element is getting returned.

File: spec/queue.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      myQueue.enqueue(99)
+      expect(myQueue.isEmpty()).to.eql(false)
+    })
+  })
+
+  context('#length', () => {
+    it('should return the number of elements in the queue', () => {
+      myQueue.enqueue("foo")
+      expect(myQueue.length()).to.eql(1)
+      expect(() => myQueue.enqueue("bar")).to.alter(() => myQueue.length(), { from: 1, to: 2 })
```
Comment (jrobcodes) [inline spec/queue.js]: This expectation tests `enqueue` - it just uses `length` to verify that `enqueue` works.  I don't think it's necessary in this context.

I would, however, add a test to ensure that an empty queue returns 0, just to make sure I test all possible cases.

File: CONTRACT.md [hunk 7]
```diff
@@ -1,251 +0,0 @@
-# Learning Contract for Goal: Core Data Structures
-
```
Comment (nicosesma) [inline CONTRACT.md:2]: Did you delete the contract???

File: src/queue.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  dequeue() {
+    let list = this.front()
+    this._list = this._list.reduce( ( memo, element, index ) => {
+      if ( this._list[ index + 1 ] !== undefined ) {
+        memo[ index ] = this._list[ index + 1 ]
+      }
+      return memo
+    }, [] )
+    return list
```
Comment (nicosesma) [inline src/queue.js:19]: This looks great, you could also even have it as
```javascript
return this._list.splice(0, 1)[0]
```

File: src/queue.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  front() {
+    return this._list[ this.length() - this._backElementPosition ] || null
+  }
```
Comment (nicosesma) [inline src/queue.js:23]: I wonder if `return this._list[0] || null` would work because it seems as if the operation that you're doing should always equal zero for the value of the array's index.

File: src/stack.js [hunk 10]
```diff
@@ -1,5 +0,0 @@
-'use strict'
-
-export default class Stack {
-  // your code here
```
Comment (nicosesma) [inline src/stack.js:4]: Are you deleting this file along with the supposed implementation's tests??

**Review by nicosesma — COMMENTED** [review]
Looks good 👍
